File: examples/ecommerce-ts/handlers/list_orders.ts
import type { RouteRequest, RouteResult, PrefetchMessage, PrefetchDb, HandleContext, RenderContext } from "focus";
import { price } from "focus";

// [route] .list_orders
// match GET /orders
export function route(_req: RouteRequest): RouteResult | null {
  return {};
}

// [prefetch] .list_orders
export async function prefetch(_msg: PrefetchMessage, db: PrefetchDb) {
  const orders = await db.queryAll("SELECT id, total_cents, items_len, status, timeout_at, payment_ref FROM orders ORDER BY id LIMIT ?1", 50);
  return { orders };
}

// [handle] .list_orders
export function handle(_ctx: HandleContext): string {
  return "ok";
}

// [render] .list_orders
export function render(ctx: RenderContext): string {
  const orders = ctx.prefetched.orders as unknown[] | null;
  if (!orders || orders.length === 0) return "<div>No orders</div>";
  return orders.map((o: any) =>
    `<div class="card">Order ${o.id.slice(0, 8)}... &mdash; ${o.status} &mdash; ${price(o.total_cents)} &mdash; ${o.items_len} items</div>`
  ).join("");
}

File: examples/ecommerce-py/handlers/list_orders.py
import sys
sys.path.insert(0, str(__import__("pathlib").Path(__file__).resolve().parents[2] / "adapters" / "python"))
from tiger_web import price


def route(req):
    return {"operation": "list_orders", "id": "0" * 32}


def prefetch(msg, db):
    orders = db.query_all(
        "SELECT id, total_cents, items_len, status, timeout_at, payment_ref FROM orders ORDER BY id LIMIT ?1", 50
    )
    return {"orders": orders}


def handle(ctx, db):
    return "ok"


def render(ctx, db=None):
    orders = ctx["prefetched"].get("orders") or []
    if not orders:
        return '<div class="meta">No orders</div>'

    return "".join(
        f'<div class="card">status={o["status"]} &mdash; {price(o["total_cents"])}'
        f' &mdash; {o["items_len"]} items'
        f'<div class="meta">{o["id"]}</div></div>'
        for o in orders
    )
